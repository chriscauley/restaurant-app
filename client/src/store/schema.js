import { reactive } from 'vue'

import api from '@/common/api'

const state = reactive({})

const fetch = form_name => {
  if (!state[form_name]) {
    api.get(`${form_name}/?schema=1`).then(r => (state[form_name] = r.schema))
  }
  return state[form_name]
}

const markStale = form_name => delete state[form_name]

export default { fetch, markStale }
