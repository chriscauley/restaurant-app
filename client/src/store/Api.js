import { reactive } from 'vue'

import api from '@/common/api'

export default () => {
  let stale_at = new Date().valueOf()
  const url_fetched_at = {}
  const loading = reactive({})
  const state = reactive({})
  const get = url => {
    const needs_fetch = stale_at > url_fetched_at[url] || !state[url]
    if (needs_fetch && !loading[url]) {
      loading[url] = true
      api.get(url).then(data => {
        url_fetched_at[url] = new Date().valueOf()
        state[url] = data
        loading[url] = false
      })
    }
    return state[url]
  }

  const markStale = () => {
    stale_at = new Date().valueOf()
  }

  const post = (url, data) => {
    api.post(url, data).then(() => {
      markStale()
      get(url)
    })
  }

  return { get, markStale, post }
}
