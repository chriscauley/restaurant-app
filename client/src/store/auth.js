import { reactive } from 'vue'
import api from '@/common/api'

const pending = []

const state = reactive({
  user: null,
  loading: null,
  loaded: null,
})

const logout = () =>
  api.post('logout').then(() => {
    state.user = null
    check()
  })

const check = async () => {
  if (state.loading) {
    return new Promise(resolve => pending.push(resolve))
  }
  if (!state.loaded) {
    state.loading = true
    return api.get('whoami').then(({ user }) => {
      state.user = user
      state.loading = false
      state.loaded = true
      while (pending.length) {
        // resolve any other pending promises
        pending.pop()()
      }
    })
  }
}

export default {
  state,
  check,
  init: check,
  logout,
}
