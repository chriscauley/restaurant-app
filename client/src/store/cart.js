import { ReactiveRestApi } from '@unrest/vue-storage'
import client from '@/common/api'
import router from '@/router'
import uS from '@unrest/story'

const complete = action => data => {
  uS.complete(`customer.${action}`)
  return fetch(data)
}

const store = ReactiveRestApi({ client })

const fetch = () => store.fetch('cart/')
const get = () => store.get('cart/')

const addItem = item_id => store.post('cart/add/', { item_id }).then(complete('createCartItem'))

const removeItem = item_id =>
  store.post('cart/remove/', { item_id }).then(complete('removeCartItem'))

const checkout = () => {
  store.post('cart/checkout/').then(({ order_id }) => {
    uS.complete('customer.placeOrder')
    router.push({ name: 'orderdetail', params: { order_id } })
  })
}

export default { get, fetch, addItem, removeItem, checkout }
