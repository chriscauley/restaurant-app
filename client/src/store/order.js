import { RestStorage } from '@unrest/vue-storage'

import client from '@/common/api'

const store = RestStorage('order', { client })

const refetch = ({ id }) => store.getOne(id)

const updateStatus = (id, status) => store.save({ id, status }).then(refetch)
const blockUser = id => store.save({ id, action: 'block' }).then(refetch)
const unblockUser = id => store.save({ id, action: 'unblock' }).then(refetch)

Object.assign(store, { updateStatus, blockUser, unblockUser })

export default store
