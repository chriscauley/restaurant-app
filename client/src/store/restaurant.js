import { RestStorage } from '@unrest/vue-storage'

import client from '@/common/api'

import { slugify } from '@/common/utils'

const DEFAULT_IMG = '/static/pizza.png'

const fromServer = data => {
  data.img_style = `background-image: url(${data.photo_url || DEFAULT_IMG})`
  data.to = `/restaurant/${data.id}/${slugify(data.name)}/`
  return data
}

export default RestStorage('restaurant', { client, fromServer })
