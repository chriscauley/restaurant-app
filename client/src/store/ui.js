import { reactive } from 'vue'

let ID_COUNTER = 0
const DELAY = 10000

const state = reactive({
  alert: null,
  toasts: [],
})

const hide = id => {
  const item = state.toasts.find(item => item.id === id)
  item.hidden = true
}

export default {
  alert: item => (state.alert = item),
  closeAlert: () => (state.alert = null),
  toast: item => {
    const id = ID_COUNTER++
    item.id = id
    item.hidden = false
    state.toasts.push(item)
    setTimeout(() => hide(item.id), DELAY)
  },
  state,
}
