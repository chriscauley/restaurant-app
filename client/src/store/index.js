import auth from './auth'
import cart from './cart'
import order from './order'
import restaurant from './restaurant'
import schema from './schema'
import ui from './ui'

const store = {
  list: [],
  install: (app, _options) => {
    app.config.globalProperties.$store = store
    store.list.forEach(m => m.init?.())
  },
}

Object.entries({ auth, cart, order, restaurant, schema, ui }).forEach(([name, module]) => {
  store.list.push(module)
  store[name] = module
})

export default store
