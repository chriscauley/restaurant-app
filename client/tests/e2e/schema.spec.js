import store from '@/store'

const forms = ['login', 'signup', 'restaurant', 'owner/signup', 'menuitem', 'menusection']

forms.forEach(form_name =>
  test('getting schema form ' + form_name, done => {
    store.schema.fetch(form_name)
    setTimeout(() => {
      expect(store.schema.fetch('form_name')).toMatchSnapshot()
      done()
    }, 1000)
  }),
)
