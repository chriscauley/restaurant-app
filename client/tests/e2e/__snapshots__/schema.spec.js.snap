// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getting schema form login 1`] = `undefined`;

exports[`getting schema form menuitem 1`] = `undefined`;

exports[`getting schema form menusection 1`] = `undefined`;

exports[`getting schema form owner/signup 1`] = `undefined`;

exports[`getting schema form restaurant 1`] = `undefined`;

exports[`getting schema form signup 1`] = `undefined`;
